### PR TITLE
Improve countersigning scenario

### PR DESCRIPTION
**Pull request by ThetaSinner**
image

### CODE REVIEW

File: bindings/trycp_runner/src/common.rs [hunk 1]
```diff
@@ -314,7 +314,7 @@ where
         let agent_name = agent_name.clone();
         async move {
             let logs = client
-                .download_logs(agent_name, Some(Duration::from_secs(180)))
+                .download_logs(agent_name, Some(Duration::from_secs(10 * 60)))
```
Comment (ThetaSinner) [inline bindings/trycp_runner/src/common.rs:317]: When the logs are outputting a lot of information it can take a few minutes to get everything

File: conductor-config-ci.yaml [hunk 2]
```diff
@@ -1,3 +1,6 @@
+dpki:
+  dna_path: ~
+  no_dpki: true
```
Comment (ThetaSinner) [inline conductor-config-ci.yaml:3]: - [ ] Needs @maackle's change to DPKI for test mode before this can be resolved

File: flake.nix [hunk 3]
```diff
@@ -8,6 +8,7 @@
     holonix = {
       url = "github:holochain/holonix?ref=main";
       inputs = {
+        holochain.url = "github:holochain/holochain?ref=4159-countersigning-workflow-rework-debug";
```
Comment (ThetaSinner) [inline flake.nix:11]: - [ ] Needs a release of Holochain that has the countersigning fixes

File: scenarios/two_party_countersigning/src/main.rs [hunk 4]
```diff
@@ -117,132 +119,109 @@ fn agent_behaviour_initiate(
...
-                            }
-                        }
+                    let session_times = my_preflight_response.request.session_times.clone();
+                    let session_timeout = Instant::now().add(Duration::from_millis(
+                        (session_times.end.as_millis() - session_times.start.as_millis()) as u64,
+                    ));
 
-                        // Add no new agents, that should only happen when we exhaust the list.
-                        Ok(Vec::with_capacity(0))
+                    match run_initiated_session(
```
Comment (ThetaSinner) [inline scenarios/two_party_countersigning/src/main.rs:192]: Hideous diff, the code to the left and above has been moved to a function. The reason is to make sure that any errors that happen after the session has started won't cause the agent behavior to continue to the next iteration. We have to wait for the session to time out if something goes wrong!

File: scenarios/two_party_countersigning/src/main.rs [hunk 5]
```diff
@@ -252,6 +231,102 @@ fn agent_behaviour_initiate(
...
+                    Signal::App { signal, .. } => signal,
+                    _ => {
+                        log::debug!("Received a signal that is not an app signal, listening for other signals.");
+                        continue;
+                    }
+                };
+
+                let other_response = match signal.clone().into_inner().decode::<Signals>() {
+                    Ok(Signals::Response(response))
+                        if session_fingerprint(response.request())?
```
Comment (ThetaSinner) [inline scenarios/two_party_countersigning/src/main.rs:265]: This changed. Previously I was comparing the `app_entry_hash` which isn't unique and was causing crossed wires and so validation failures. This is how Holochain checks that sessions are equal and massively improves the scenario behavior.

File: scenarios/two_party_countersigning/src/main.rs [hunk 6]
```diff
@@ -488,6 +563,15 @@ async fn complete_session(
     Ok(retry_count)
 }
 
+fn session_fingerprint(preflight_request: &PreflightRequest) -> anyhow::Result<Vec<u8>> {
+    let hash = holo_hash::encode::blake2b_256(
```
Comment (ThetaSinner) [inline scenarios/two_party_countersigning/src/main.rs:567]: - [ ] Move to Holochain, it can be a function on the `PreflightRequest`

File: scenarios/two_party_countersigning/src/main.rs [hunk 7]
```diff
@@ -117,132 +119,109 @@ fn agent_behaviour_initiate(
...
+                        Ok(_) => {
+                            // Completed successfully
+                        }
+                        Err(e) => {
+                            log::warn!(
+                                "Failed to run initiated session, waiting for it to time out: {:?}",
+                                e
+                            );
+
+                            // TODO this can be replaced with listening for the abandoned signal
```
Comment (ThetaSinner) [inline scenarios/two_party_countersigning/src/main.rs:215]: - [ ] Timeout signals need to be handled in a couple of places through this test. Update this once Holochain is released with that signal defined

**Review by neonphog — DISMISSED** [review]
lgtm : )

File: scenarios/two_party_countersigning/src/main.rs [hunk 8]
```diff
@@ -507,10 +614,20 @@ async fn await_countersigning_success(
...
                     }
+                    Signal::System(SystemSignal::AbandonedCountersigning(eh))
+                        if eh == session_entry_hash =>
+                    {
+                        return Err(anyhow::anyhow!("Countersigning session was abandoned"));
+                    }
+                    Signal::System(SystemSignal::AbandonedCountersigning(_)) => {
+                        // This shouldn't happen because only one countersigning session can be active at a time. There's a bug if this log message shows up.
+                        log::warn!("Received an abandoned countersigning signal for a different session, listening for other signals.");
+                        continue;
```
Comment (cdunster) [inline scenarios/two_party_countersigning/src/main.rs:629]: Given the comment then should we log this as an error instead?

File: scenarios/two_party_countersigning/Cargo.toml [hunk 9]
```diff
@@ -12,6 +12,8 @@ log = { workspace = true }
 rmp-serde = { workspace = true }
 
 holochain_types = { workspace = true }
+holo_hash = { workspace = true }
+holochain_serialized_bytes = { workspace = true }
```
Comment (cdunster) [inline scenarios/two_party_countersigning/Cargo.toml:16]: Are these dependencies used? I see them added here but I don't see a `use` statement anywhere.

File: conductor-config-ci.yaml [hunk 10]
```diff
@@ -1,6 +1,21 @@
+#danger_generate_throwaway_device_seed: true
+#dpki:
+#  dna_path: ~
+#  network_seed: wind-tunnel-test
+#  allow_throwaway_random_dpki_agent_key: true
+#  no_dpki: false
```
[the remark this answers is not in the document]
Comment (ThetaSinner) [reply conductor-config-ci.yaml:6]: Ideally not removed no. I'd prefer that we can test either way. There's a conversation on #92 about ways to do that.

For now, enabling DPKI would require changing code to pass the config through to common helper functions so they can modify their behaviour around app installs. You can't provide an agent key if you're using DPKI. I'm also not sure that's entirely intentional behaviour around that, but that's another issue.

Basically, more work required yet to support testing with either config.

File: scenarios/two_party_countersigning/src/main.rs (hunk 8 above)
Comment (ThetaSinner) [reply scenarios/two_party_countersigning/src/main.rs:629]: Fair point, yes

File: scenarios/two_party_countersigning/src/main.rs [hunk 11]
```diff
@@ -315,45 +390,48 @@ fn agent_behaviour_participate(
...
-                        ).await?;
-
-                        log::debug!("Accepted the incoming session, proceeding to commit.");
+                        match run_accepted_session(client, request, session_timeout, cell_id, app_port).await {
+                            Ok(retry_count) => {
+                                let elapsed = start.elapsed();
+
+                                log::debug!("Completed countersigning session with the initiating party.");
+
+                                let accepted_success = accepted_success.fetch_add(1, std::sync::atomic::Ordering::Acquire);
```
[the remark this answers is not in the document]
Comment (ThetaSinner) [reply scenarios/two_party_countersigning/src/main.rs:399]: Yep, this is multi-threaded but each agent runs on a single thread and there's no actual parallelism within that agent's behaviour. Relaxed sounds sensible. I've updated all uses here

File: scenarios/two_party_countersigning/Cargo.toml (hunk 9 above)
Comment (ThetaSinner) [reply scenarios/two_party_countersigning/Cargo.toml:16]: Nope, removed

File: conductor-config-ci.yaml (hunk 10 above)
[the remark this answers is not in the document]
Comment (jost-s) [reply conductor-config-ci.yaml:6]: > You can't provide an agent key if you're using DPKI. I'm also not sure that's entirely intentional behaviour around that, but that's another issue.

I agree, that's a confusing behavior that I've also noticed. It will need to be possible for Holo anyway and other people who generate their keys outside of Holochain app installation.

**Review by neonphog — APPROVED** [review]
